presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,225 @@
+
+    private static ExecutorService executor;
+
+    @Inject
+    @Named("databases.presto.server_address")
+    private String serverAddress;
+
+    @BeforeTestWithContext
+    @AfterTestWithContext
+    public void dropTestTables()
+    {
+        try {
+            query(format("DROP TABLE IF EXISTS %s", tableName));
+        }
+        catch (Exception e) {
+            Logger.get(getClass()).warn(e, "failed to drop table");
+        }
+    }
+
+    @BeforeTestWithContext
+    public void startExecutor()
+    {
+        executor = Executors.newFixedThreadPool(2);
+    }
+
+    @AfterTestWithContext
+    public void stopExecutor()
+    {
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(60, TimeUnit.SECONDS)) {
+                executor.shutdownNow();
+            }
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("failed to stop the executor", e);
+        }
+    }
+
+    private class SqlExecuteCallable
+            implements Callable<Boolean>
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Boolean call()
+        {
+            try {
+                query(sql);
+                return false;
+            }
+            catch (QueryExecutionException ExceptionDueToQueryCancellation) {
+                return true;
+            }
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+        private long waitTime;
+        private long step;
+
+        public SqlCancelCallable(String sqlStatement, long waitTime, long step)
+        {
+            this.sqlStatement = sqlStatement;
+            this.waitTime = waitTime;
+            this.step = step;
+        }
+
+        public Response call()
+        {
+            long time = 0;
+            while (time < waitTime) {
+                List<Object> queryResult = query(format("SELECT state, query_id from system.runtime.queries" +
+                        " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement)).row(0);
+                String queryState = (String) queryResult.get(0);
+                String queryId = (String) queryResult.get(1);
+                if (!queryState.equalsIgnoreCase("running")) {
+                    try {
+                        Thread.sleep(step);
+                    }
+                    catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        Logger.get(getClass()).warn(e, "error in SqlCancelCallable");
+                        return null;
+                    }
+                    time = time + step;
+                }
+                else {
+                    return (new QueryCanceller(queryId, serverAddress)).cancel();
```

Comment:
Instead of creating a new QueryCanceller every time, create one where you create the executor, and when you call the cancel command, pass in the queryId